### PR TITLE
Storage+REST supporting `follower` mode.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Storage/test.cc [hunk 1]
```diff
@@ -934,22 +937,41 @@ TEST(TransactionalStorage, ReplicationViaHTTP) {
...
+  const auto replicated_stream_file_remover = current::FileSystem::ScopedRmFile(replicated_stream_file_name);
+  using transaction_t = typename Storage::transaction_t;
+  using sherlock_t = current::sherlock::Stream<transaction_t, current::persistence::File>;
+  sherlock_t replicated_stream(replicated_stream_file_name);
+
+  struct StreamPublisherOwner {
+    sherlock_t& stream_;
+    std::unique_ptr<sherlock_t::publisher_t> publisher_;
+    explicit StreamPublisherOwner(sherlock_t& stream) : stream_(stream) {}
+    ~StreamPublisherOwner() { assert(!publisher_); }
```
Comment (dkorolev) [inline Storage/test.cc]: s/assert/ASSERT_FALSE(static_cast<bool>(publisher_))/

File: Storage/transaction_policy.h [hunk 2]
```diff
@@ -96,9 +98,47 @@ class Synchronous final {
     return future;
   }
 
+  // Read-only transaction returning non-void type.
+  template <typename F, class = ENABLE_IF<!std::is_void<f_result_t<F>>::value>>
```
Comment (dkorolev) [inline Storage/transaction_policy.h]: `s/ENABLE_IF/enable_if_t/` ?

File: Storage/transaction_policy.h [hunk 3]
```diff
@@ -131,11 +171,45 @@ class Synchronous final {
     return future;
   }
 
+  // Read-only transaction returning void type.
+  template <typename F, class = ENABLE_IF<std::is_void<f_result_t<F>>::value>>
```
Comment (dkorolev) [inline Storage/transaction_policy.h]: `s/ENABLE_IF/enable_if_t/` ?

File: Storage/transaction_policy.h [hunk 4]
```diff
@@ -172,17 +246,43 @@ class Synchronous final {
...
   }
 
-  template <typename F>
-  void ReplayTransaction(F&& f, transaction_t&& transaction, current::ss::IndexAndTimestamp idx_ts) {
-    std::lock_guard<std::mutex> lock(mutex_);
-    persister_.ReplayTransaction(std::forward<transaction_t>(transaction), idx_ts);
-    for (auto&& mutation : transaction.mutations) {
-      f(std::move(mutation));
+  // Read-only two-step transaction.
+  template <typename F1, typename F2, class = ENABLE_IF<!std::is_void<f_result_t<F1>>::value>>
```
Comment (dkorolev) [inline Storage/transaction_policy.h]: `s/ENABLE_IF/enable_if_t/` ?

File: Storage/test.cc [hunk 5]
```diff
@@ -1439,11 +1474,172 @@ TEST(TransactionalStorage, UseExternallyProvidedSherlockStreamOfBroaderType) {
...
+  EXPECT_EQ(current::FileSystem::ReadFileAsString(master_file_name),
+            current::FileSystem::ReadFileAsString(follower_file_name));
+
+  // `FlipToMaster()` method on a storage with `Master` role throws an exception.
+  EXPECT_THROW(master_storage.FlipToMaster(), current::storage::StorageIsAlreadyMasterException);
+  // Publisher of the `follower_stream` is still in the `replicator`.
+  EXPECT_THROW(follower_storage.FlipToMaster(),
+               current::storage::UnderlyingStreamHasExternalDataAuthorityException);
+
+  // Stop replication process.
```
Comment (dkorolev) [inline Storage/test.cc]: Stop the replication process.

File: Storage/docu/docu_3_code.cc [hunk 6]
```diff
@@ -423,11 +423,16 @@ struct RESTWithMeta : current::storage::rest::AdvancedHypermedia {
   struct RESTful: SUPER::RESTful<HTTP_VERB, PARTICULAR_FIELD, ENTRY, KEY> {
     using ACTUAL_SUPER = SUPER::RESTful<HTTP_VERB, PARTICULAR_FIELD, ENTRY, KEY>;
 
-    template <class INPUT>
+    template <class INPUT, std::enable_if_t<!INPUT::read_only_transaction>* = nullptr>
```
Comment (dkorolev) [inline Storage/docu/docu_3_code.cc]: Best to put the star into `<>`, here and below.

File: Storage/api_base.h [hunk 7]
```diff
@@ -97,16 +97,20 @@ struct RESTfulRegisterTopLevelInput : RESTfulGenericInput<STORAGE> {
...
 
-template <typename STORAGE, typename FIELD>
+template <typename STORAGE, typename FIELD, bool READ_ONLY_TRANSACTION>
 struct RESTfulGETInput : RESTfulGenericInput<STORAGE> {
   using mutable_fields_t = MutableFields<STORAGE>;
-  mutable_fields_t fields;
+  using immutable_fields_t = ImmutableFields<STORAGE>;
+  using fields_type_t =
+      typename std::conditional<READ_ONLY_TRANSACTION, immutable_fields_t, mutable_fields_t>::type;
+  constexpr static bool read_only_transaction = READ_ONLY_TRANSACTION;
```
Comment (dkorolev) [inline Storage/api_base.h]: When used in `enable_if_t`, it's not immediately clear to me whether this `read_only_transaction` is a bool. Perhaps go the old way of making it a constexpr function (`IsReadOnly() { return true; }`)? The alternative is the syntax that involves `::value` in templated `enable_if_t<*read_only::value>` use cases.

**Comment by mzhurovich** [comment]
@dkorolev PTAL a8acc2fe18fc7db66bf6d74c3057ae9e85ce4941

CC @grixa Storage was refactored, `Transaction` method was split into separate `ReadOnlyTransaction` and `ReadWriteTransaction`.

**Comment by dkorolev** [comment]
@mzhurovich LGTM && merging.
CC @grixa -- will get to your code in ~10 hours.

Thanks,
Dima

**Pull request merged**